mixer/pkg/adapter/instanceUtil.go

```diff
@@ -0,0 +1,196 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package adapter
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// This file contains common utility functions that adapters need to process content
+// of instances.
+
+// Stringify converts basic data types, supported by `istio.mixer.v1.config.descriptor.ValueType`, into string.
+// Note nil object is converted into empty "" string.
+func Stringify(v interface{}) string {
+	if v != nil {
+		// Switch on all possible ValueTypes
+		switch vv := v.(type) {
+		case string:
+			return vv
+		case int64:
+			return strconv.FormatInt(vv, 10)
+		case float64:
+			return strconv.FormatFloat(vv, 'f', -1, 64)
+		case bool:
+			return strconv.FormatBool(vv)
+		case time.Time:
+			return string(serializeTime(vv))
+		case time.Duration:
+			return vv.String()
+		case net.IP:
+			return vv.String()
+		case EmailAddress:
+			return string(vv)
+		case URI:
+			return string(vv)
+		case DNSName:
+			return string(vv)
+		case map[string]string:
+			var buffer bytes.Buffer
+			for k, v := range vv {
```

Comment:
This gives a non-deterministic result since map iteration order is non-deterministic. You'll need to accumulate the keys in a slice, sort the slice, and iterate the slice.

Also, you can use pool.GetBuffer here to avoid allocating the byte.Buffer

Reply:
Do we care if this is non-deterministic? Only during StringEquals we need to ensure the comparison is order agnostic, which we are doing. This string is just to saving in DB or printing. I have a test for this https://github.com/istio/istio/pull/3391/files/48bb5457e0a1775cfb50109b53a8a00e062e6097#diff-4f7b39f5e9706d4bc1c336abea02aecfR86

Will use the pool.

Reply:
I think I'd feel better about deterministic output, since we don't know the scenario where the string will actually be used. It's unlikely to matter, but just for safety's sake, I think it'd be better to sort.

Reply:
Regarding deterministic output, I was mainly concerned about the extra work of sorting during Stringify, since it is going to be commonly used by adapters. Well, I guess, safety first than performance.